font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,157 @@
 // limitations under the License.
 //
 //! Rasterization powered by DirectWrite
-use dwrote::{FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::os::windows::ffi::OsStringExt;
+
+use dwrote::{
+    Font, FontCollection, FontFace, FontFallback, FontStretch, FontStyle, FontWeight, GlyphOffset,
+    GlyphRunAnalysis
+};
+
+use winapi::shared::ntdef::LOCALE_NAME_MAX_LENGTH;
+use winapi::um::dwrite;
+use winapi::um::winnls::GetUserDefaultLocaleName;
 
 use super::{
     BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight,
 };
 
+pub struct LoadedFont {
+    font: Font,
+    font_face: FontFace,
+}
+
 pub struct DirectWriteRasterizer {
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, LoadedFont>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    font_collection: FontCollection,
+    font_fallback: Option<FontFallback>,
 }
 
-impl crate::Rasterize for DirectWriteRasterizer {
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        let loaded_font = &self.fonts.get(&glyph.font_key).ok_or(Error::FontNotLoaded)?;
 
-    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
-        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+        match self.rasterize_font_char(&loaded_font.font_face, glyph.c, glyph.size) {
+            Ok(glyph) => Ok(glyph),
+            Err(Error::MissingGlyph(_)) => {
+                let fallback_font = self.get_fallback_font(&loaded_font.font, glyph.c)?;
+                self.rasterize_font_char(&fallback_font.create_font_face(), glyph.c, glyph.size)
+            },
+            Err(err) => Err(err),
+        }
     }
 
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn get_fallback_font(&self, font: &Font, c: char) -> Result<Font, Error> {
+        let fallback = match self.font_fallback.as_ref() {
+            Some(fallback) => fallback,
+            None => return Err(Error::MissingGlyph(c)),
+        };
+
+        let utf16_codepoints = encode_utf16(c);
+        let length = utf16_codepoints.len() as u32;
+
+        let locale = get_current_locale();
+        let text_analysis_source = dwrote::TextAnalysisSource::from_text(
```

Comment:
```suggestion
        let text_analysis_source = TextAnalysisSource::from_text(
```

Should probably `use dwrote::TextAnalysisSource` at the top.